### DISCUSSION TITLE
Filter taskbar-wheel-cycle listing

### DISCUSSION THREAD

**Discussion by Skylli202** [ramensoftware/windhawk-mods#482] [Q&A]
Hi,

I'd like to ask guideline, or general help on how should proceed to fork and edit the mode `taskbar-wheel-cycle`. I'm a SWE myself, with some 5yo C++ knowledge that is a bit dusty, but I do a decent amount of Rust, so I should be able to code it myself. I just want to be pointed into the correct direct, if my goal is achievable. Alright...

I do not take advantage of the scroll wheel to switch tab. I am way more interested in the `alt + [` and `alt +]` key binds to cycle through the taskbar. I'd like to slightly customize the behavior, but maybe it could be an option in the standard mod. Let me explain.

In my usage, I have an old piece of software (2004/2005 desktop app). And I open a lot of instances (8 to 10 of it) because it does not have tabs like a web browser have. I use the mod `Disable grouping on the taskbar` with the grouping mod setting inverted (only disable grouping if excluded) and I exclude my `software.exe` executable name.

So, how could I limit the `taskbar-wheel-cycle` mod to specific executable, like the `Disable grouping on the taskbar` is. So, I can cycle through my application instance instead of the whole task bar?

----

For the full story:

I have made a very low quality C++ program (one of those atrocious ugly one file program, yes I'm ashamed of myself) that were taking advantage of the `windows.h`, `windowsx.h` and `winuser.h`:
- `EnumWindows` to find all the instance of my app, finding in the taskbar is way smarter, and store them in a vector.
- `GetwindowText` to check if it was the application I was looking for.
- As I had a lot of issue finding how to bring back my app to foreground regardless if it was minimized, or not, and have focus in it. I just threw all the following at it: `ShowWindowAsync`, `SetForeGroundWindow`, `SetActiveWindow`, `SetFocus` and `BringWindowToTop`.
- Add some `RegisterHotKey` to have my own key binds (next and previous app, refresh app list, etc.

I was using it with 7+ Task bar. I am now running Windows 11.

Thanks in advance

**Comment by m417z** [comment]
It can certainly be done with some amount of work. Here you can see how minimized windows are skipped:
https://github.com/ramensoftware/windhawk-mods/blob/a61b96c8ecb6a86ce4c3af434ab3bade823b67a2/mods/taskbar-wheel-cycle.wh.cpp#L341-L342

Similarly, you can skip windows depending on other criteria such as the process name. The more challenging part would be to retrieve the process name from the data that you have.

I think the easiest way is to rely on the AppId. You can use the iterated button group: `button_groups[button_group_index]`.
With the button group, you can get the task group as following:
https://github.com/ramensoftware/windhawk-mods/blob/a61b96c8ecb6a86ce4c3af434ab3bade823b67a2/mods/taskbar-grouping.wh.cpp#L752
Then you can get the AppId from it as following:
https://github.com/ramensoftware/windhawk-mods/blob/a61b96c8ecb6a86ce4c3af434ab3bade823b67a2/mods/taskbar-grouping.wh.cpp#L797
It probably seems scary for somebody who's new to this code. Indeed, even simple things can be difficult since all this stuff is undocumented. Let me know if you need help with this.

**Reply by Skylli202** [reply]
Thank you very much for your (very) quick answer and availability. I'll try to have a look whenever I can.

**Reply by Skylli202** [reply]
I did not see the tab "Advanced" in the detail of the mod, before now. And in there, there is some option that seems interesting, but I might not be too much familiar with the wording. 

![WFLRfOQGUj](https://github.com/ramensoftware/windhawk-mods/assets/38432216/db84c186-d905-485c-bb1d-4aa013cd72d2)

I tried to enable the ignore switch, as you can see on the screenshot. And add my application.exe (the same values that I used on the ungrouping mod) in the custom lists. Should it produce the expected result? Seems that I'm off by something because I now scroll between nothing when I use the key binds (or the mouse wheel).

**Reply by m417z** [reply]
The custom lists control which programs the mod affects. In this case, the mod affects only one program, explorer.exe which controls the taskbar. In other words, the lists aren’t relevant for this mod. It's more relevant for mods which are designed to work with multiple or all programs, such as [Slick Window Arrangement](https://windhawk.net/mods/slick-window-arrangement).

**Reply by Skylli202** [reply]
I see. Thanks for the knowledge share.
